test/base/utils/TestITensorsExportedNames/TestITensorsExportedNames.jl

```diff
@@ -181,167 +271,83 @@ const ITENSORS_EXPORTED_NAMES = [
   :loginner,
   :lognorm,
   :lproj,
-  :lq,
-  :mapprime,
-  :mapprime!,
-  :matmul,
-  :matrix,
   :maxdim,
   :maxdim!,
   :maxlinkdim,
   :measure!,
   :measurements,
   :mindim,
   :mindim!,
-  :modulus,
   :movesite,
   :movesites,
   :mul,
-  :mul!,
   :multMPO,
   :multmpo,
-  :nblocks,
-  :nnz,
-  :nnzblocks,
   :noise,
   :noise!,
   :noiseterm,
-  :noncommonind,
-  :noncommoninds,
-  :noprime,
-  :noprime!,
-  :norm,
-  :normalize,
-  :normalize!,
-  :not,
   :nsite,
   :nsweep,
-  :nullspace,
-  :nzblock,
-  :nzblocks,
-  :onehot,
   :op,
   :ops,
-  :order,
   :orthoCenter,
   :ortho_lims,
   :orthocenter,
   :orthogonalize,
   :orthogonalize!,
   :outer,
-  :permute,
-  :plev,
-  :polar,
-  :pop,
-  :popfirst,
   :position!,
-  :prime,
-  :prime!,
-  :primelinks!,
   :product,
+  :primelinks!,
   :productMPS,
   :projector,
   :promote_itensor_eltype,
-  :push,
-  :pushfirst,
-  :ql,
-  :qn,
-  :qr,
-  :randn!,
   :randomITensor,
   :randomMPO,
   :randomMPS,
-  :readcpp,
-  :removeqn,
-  :removeqns,
-  :removetags,
-  :removetags!,
   :replace_siteinds,
   :replace_siteinds!,
   :replacebond,
   :replacebond!,
-  :replaceind,
-  :replaceind!,
-  :replaceindex!,
-  :replaceinds,
-  :replaceinds!,
   :replaceprime,
   :replacesites!,
-  :replacetags,
-  :replacetags!,
   :reset_ortho_lims!,
-  :reset_warn_order!,
-  :rmul!,
   :rproj,
-  :rq,
   :sample,
   :sample!,
-  :scalar,
-  :scale!,
   :set_leftlim!,
   :set_ortho_lims!,
   :set_rightlim!,
-  :set_warn_order!,
   :setcutoff!,
-  :setdir,
-  :setelt,
-  :setindex,
   :setmaxdim!,
   :setmindim!,
   :setnoise!,
-  :setprime,
-  :setprime!,
-  :setspace,
-  :settags,
-  :settags!,
-  :sim,
-  :sim!,
   :simlinks!,
   :siteind,
   :siteindex,
   :siteinds,
-  :space,
   :splitblocks,
   :square_lattice,
   :state,
-  :storage,
-  :store,
   :sum,
-  :svd,
   :swapbondsites,
-  :swapind,
-  :swapinds,
-  :swapinds!,
-  :swapprime,
-  :swapprime!,
-  :swaptags,
-  :swaptags!,
   :sweepnext,
-  :tags,
   :tensors,
   :toMPO,
   :totalqn,
   :tr,
-  :transpose,
   :triangular_lattice,
   :truncate,
   :truncate!,
   :truncerror,
   :truncerrors,
-  :unionind,
-  :unioninds,
   :unique_siteind,
   :unique_siteinds,
-  :uniqueind,
-  :uniqueindex,
-  :uniqueinds,
-  :use_combine_contract,
-  :use_debug_checks,
   :val,
```

Comment:
Similarly, I would say the above list of `ITensorMPS` exported names all look to be correctly MPS related, and not needed for `ITensors.jl` alone, with maybe these exceptions:
* `state` 
* `siteind`
* `siteinds` maybe also if `siteind` goes back into ITensors.jl?
* similarly `op` and related functions

Ah but I see, you have moved these into a `lib/` with an extension (I saw the PR about that but only just remembered). So then is the idea of moving these exports to ITensorMPS more about whether they are exported by default (as ITensorMPS would) versus making them optional (as ITensors.jl now would)?

Reply:
Yes, that's a good summary. Glad to see you were able to determine the logic behind these decisions without an explanation, that's a good sanity check!

Yes, the idea would be that we could provide some alternative for loading functionality like `state`, `siteind[s]`, `op`, etc. instead of exporting them by default from ITensors.jl. The initial suggestion will be to load `ITensorMPS.jl` (which is not much worse than the current situation anyway, in terms of burden of how much code gets loaded/compiled/etc.). In the future the goal will be to split off that functionality into separate packages that are either independent of or built on top of ITensors.jl.

For example, I could imagine something like:
- `LocalHilbertSpaces.jl`: Definitions like `SiteType`, `siteind`, etc. (maybe with one layer independent of ITensor that builds spaces and another that builds `Index` objects).
- `QuantumStates.jl`: Local state definitions (up, down, etc.) and things like `StateName`, independent of ITensor (say outputs dense or sparse vector representions of states or functions for setting elements independent of the particular data structures).
- `QuantumOperators.jl`: Local operators definitions (Sx, Sy, etc.) and things like `OpName`, independent of ITensor (say outputs dense or sparse matrix representions of operators or functions for setting elements independent of the particular data structures). Maybe combine with `QuantumStates.jl`.
- `ITensorStates.jl` and `ITensorOperators.jl`: Compatiblity layer between `QuantumStates.jl` and `QuantumOperators.jl` with the particular task of converting state and operator definitions into ITensor representations.
 
Obviously the particular modular structure and names will need to be discussed, I'm just giving a brief outline of a potential design.

So we could picture that these packages, along with `LazyApply` and `Ops`, sit between ITensors.jl and ITensorMPS.jl/ITensorNetworks.jl. But for the time being while we work out all of those details we can give users the instructions that they should load `ITensors` and `ITensorMPS` if they want customizable Hilbert space, state, and operator construction functionality. The only downside of that is that users that want Hilbert space, state, and operator definitions may find it strange that they have to load a full MPS package to get that, but that is already what we make them do with ITensors.jl so I don't think it is too bad. The benefit for us is that it will give us a lot more flexibility to break apart ITensors.jl into smaller composable pieces in the medium to long run while imposing fewer breaking changes for users during that process.